### ISSUE TITLE
Setting up Azure pipelines

### ISSUE THREAD

**Issue by aminya** [atom-ide-community/atom#1]
We need to set up Azure pipelines to get CI similar to upstream.

**Comment by aminya** [comment]
In this repository, we use projects as our backlog and a way to track the changes we make. This will help us allowing the enhancements to be considered for the upstream repository, in case they were interested.

**Comment by DeeDeeG** [comment]
Checking in, that all sounds good. :+1:

**Comment by aminya** [comment]
@DeeDeeG Could you set up the CI? I saw you had Azure Pipelines running on your personal account. How did you do this?

**Comment by DeeDeeG** [comment]
Someone at upstream was kind enough to put my branches up at upstream and run it themselves. I'll look into if there is a free tier of Azure Pipelines though, so we can set it up here without paying some sort of sobscription.

**Comment by DeeDeeG** [comment]
For the email/notifications record: 

> Edit: This looks like the place to do it. "Start free with GitHub" https://azure.microsoft.com/en-us/services/devops/pipelines/

> _Kinda makes sense now that Microsoft owns Github..._

**Comment by aminya** [comment]
> Someone at upstream was kind enough to put my branches up at upstream and run it themselves. I'll look into if there is a free tier of Azure Pipelines though, so we can set it up here without paying some sort of subscription.
> 
> Edit: This looks like the place to do it. "Start free with GitHub" https://azure.microsoft.com/en-us/services/devops/pipelines/
> 
> _Kinda makes sense now that Microsoft owns Github..._

There is a free tier, but how do we get the same configuration that upstream is using? Is there a duplicate button or something, or is there a config file we need to use somewhere? I could not find anything in the repository itself.

**Comment by DeeDeeG** [comment]
Ah yeah, the config is in a pretty obscure place. I only found it because it mentions Python, and I did some PRs about being ready for Python3.

https://github.com/atom/atom/tree/master/script/vsts

(_VSTS_ is short for _Visual Studio Team System_, which got renamed to _Azure Devops Server_... according to [Wikipedia](Azure_DevOps_Server).)

**Comment by DeeDeeG** [comment]
Particularly here: https://github.com/atom/atom/tree/master/script/vsts/platforms

**Comment by DeeDeeG** [comment]
Editing this as I go to be as complete as I can manage:

- Visit here https://azure.microsoft.com/services/devops/pipelines/ and press "Start free with GitHub"
- Visit https://dev.azure.com
- Click the "+ New project" button (it's blue for me)
  - Fill in some details (name, description) and press "Create"
- In the new project, click "Pipelines" in the side nav area
  - It asks you "Where is your code"
    - Choose this repo `atom-ide-community:atom`
    - Authenticate Azure Pipelines to GitHub via OAuth
- ??? (I don't have owner rights for the `atom-ide-community` (which is fine by me for the moment), so I can't see what the next steps are from here.)

**Comment by aminya** [comment]
I have created the repository: https://dev.azure.com/atomcommunity/atomcommunity
I am now trying to see how I can set up build using the files in https://github.com/atom/atom/tree/master/script/vsts?rgh-link-date=2020-07-03T01%3A09%3A32Z 🤔

**Comment by DeeDeeG** [comment]
I'm curious what happens if you add a pipeline from the side nav bar area, and after authenticating to GitHub with OAuth so it can officially link up with/access this repo. (Trying to write out the instructions in my above comment, but that's where I get stuck.)

**Comment by DeeDeeG** [comment]
I'm trying it on my own personal fork now to see how far I can get and update my instructions/steps above.

**Comment by aminya** [comment]
I added pull requests and release builds for now: https://dev.azure.com/atomcommunity/atomcommunity/_build.
We might have to edit the release configuration once we wanted to release things with the new name.

**Comment by DeeDeeG** [comment]
I'm a bit confused about what this means, but there are "Missing tasks" required to run the CI, and supposedly these are installable via https://marketplace.visualstudio.com/ according to the error message.

**Comment by DeeDeeG** [comment]
Hmm, maybe we need to get some VMs  running? In the Pipelines sidebar there is a subsection "Environments", which is empty by default.

**Comment by DeeDeeG** [comment]
I created the "starter pipeline" with no errors.

<details><summary> Full starter pipeline yaml (click to expand)</summary>

```yml
# Starter pipeline
# Start with a minimal pipeline that you can customize to build and deploy your code.
# Add steps that build, run tests, deploy, and more:
# https://aka.ms/yaml

trigger:
- master

pool:
  vmImage: 'ubuntu-latest'

steps:
- script: echo Hello, world!
  displayName: 'Run a one-line script'

- script: |
    echo Add other tasks to build, test, and deploy your project.
    echo See https://aka.ms/yaml
  displayName: 'Run a multi-line script'
```

</details>

I can't tell if it's stalled waiting on an Azure Pipelines worker/VM, or because there are no actual steps in the script, so it never finishes... https://dev.azure.com/DeeDeeG/b/_build/results?buildId=3&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9

Now to read the docs and see if I can make one that actually does something. Working toward eventually running the Linux/Windows/macOS tests from upstream Atom.

**Comment by aminya** [comment]
Did you see #3? I need to look into the error message

**Comment by DeeDeeG** [comment]
I think this needs to be installed to the Azure DevOps org: https://marketplace.visualstudio.com/items?itemName=1ESLighthouseEng.PipelineArtifactCaching

It's referenced in all three platforms' pipeline (CI) steps.

**Comment by DeeDeeG** [comment]
Should this issue be re-purposed to "Get the CI running", or should we move discussion of the CI to another place? (I lean toward re-purposing this issue to be about getting CI working.)

Sorry about that.

**Comment by DeeDeeG** [comment]
Now I'm getting `The pipeline must contain at least one job with no dependencies`

**Comment by DeeDeeG** [comment]
I started the "Pull Requests" pipeline (`scripts/vsts/pull-requests.yml`) and it seems to be running now! (Had to install that [Lighthouse artifact caching thing](https://marketplace.visualstudio.com/items?itemName=1ESLighthouseEng.PipelineArtifactCaching) first.

Running CI on my personal fork: https://dev.azure.com/DeeDeeG/b/_build/results?buildId=7&view=results

**Comment by DeeDeeG** [comment]
Now I'm getting an `node-gyp rebuild` error for `@atom/watcher`.

Good news: CI is basically up and running for me on my fork.

I'm not sure why that package doesn't build at my fork at the moment, but it's still progress, I suppose.

**Comment by aminya** [comment]
Using Azure is pretty cool. It allows directly editing the code on its website. I was wondering if we use the same setup and a docker image or something to streamline development. Building Atom locally takes time.

**Comment by DeeDeeG** [comment]
I am kind of reluctant to use a lot of platforms and CI stuff, because I like to verify what I'm seeing locally. If you can get what you describe working, I don't see why not to try it though. :shrug:

**Comment by aminya** [comment]
Uploading the repo after bootstrap allows fast development. Doing it locally is hard or takes time. For example, on Windows, bootstrapping requires the installation of Visual Studio 15, which is pretty old.

**Comment by DeeDeeG** [comment]
I'm not sure where would let us upload that many files? The bootstrapped repo can be multiple gigabytes. And it would be a bit slow to download it. (Would it just be a zip/tarball hosted somewhere?)

Also, keep in mind that there are some native packages, so at minimum it would be one version of the bootstrapped repo per OS, maybe more than that if other stuff in the platform changes within a given OS.

So I have trouble picturing it, but I am open to seeing it happen.

**Comment by DeeDeeG** [comment]
> on Windows, bootstrapping requires a installation of Visual Studio 15, which is not desirable in 2020.

This is no longer true now that the project's `node-gyp` dependency has been updated. It works with VS 2017 (C++ build tools or full Visual Studio) and possibly even Visual Studio 2019.

The only thing needing Visual Studio is `node-gyp` which we/Atom usually bring in via depending on `npm` in a `package.json` somewhere.

The `Hacking on Atom Core` docs are a bit out of date about this. I have been meaning to confirm what info is accurate so I can update that. But I think this is accurate if you navigate to the `README.md` as of the version of `node-gyp` you are using: https://github.com/nodejs/node-gyp#on-windows

**Comment by aminya** [comment]
We need to change the bootstrap script a bit. We don't need to include the V8 image for example. That can be downloaded by the user. For the reset, when 7ziped the size will reduce. On my system the 7zip archive that has node_modules for both the root and the script folder is 452MB. We can upload the bootstrap of the current master branch every time. So in practice, we just need 3*500MB=1.5GB of space for uploading all the stuff. We can just upload those to GitHub.

**Comment by aminya** [comment]
> > on Windows, bootstrapping requires a installation of Visual Studio 15, which is not desirable in 2020.
> 
> This is no longer true now that the project's `node-gyp` dependency has been updated. It works with VS 2017 (C++ build tools or full Visual Studio) and possibly even Visual Studio 2019.
> 
> The only thing needing Visual Studio is `node-gyp` which we/Atom usually bring in via depending on `npm` in a `package.json` somewhere. (`script/package.json` and `apm/package-lock.json`).
> 
> The `Hacking on Atom Core` docs are a bit out of date about this. I have been meaning to confirm what info is accurate so I can update that. But I think this is accurate if you navigate to the `README.md` as of the version of `node-gyp` you are using: https://github.com/nodejs/node-gyp#on-windows

On my system, 2019 preview does not work. I have not tried 2017 though. I usually use WSL instead.

**Comment by DeeDeeG** [comment]
Updating the docs is In-progress: https://github.com/atom/flight-manual.atom.io/pull/630

**Comment by DeeDeeG** [comment]
Regarding standing up a release pipeline:

We would have to [pay for a `packagecloud.io` account](https://packagecloud.io/pricing), or so something else. I suggest re-write the code/write our own release pipeline to make it convenient for us to post our built Atom installers/packages somewhere other than Packagecloud.io

I vote for finding some other way to host the packages. I suppose we could distribute via personal repositories (PPA on Ubuntu, AUR on Arch, etc. etc.) and otherwise just post the installers/packages on GitHub Releases pages.

---

More about packagecloud.io if we wanted to:

If we had our own Packagecloud.io account, then (I think) we would just need to change some URLs in these files to point to it and reference this repo rather than upstream Atom:

```console
$ git grep --name-only https -- script/vsts/
script/vsts/README.md
script/vsts/get-release-version.js
script/vsts/lib/release-notes.js
script/vsts/lib/upload-linux-packages.js
script/vsts/package-lock.json
script/vsts/platforms/linux.yml
script/vsts/windows-run.js
```

**Comment by aminya** [comment]
> Regarding standing up a release pipeline:
> 
> We would have to [pay for a `packagecloud.io` account](https://packagecloud.io/pricing), or do something else. I suggest we re-write the code/write our own release pipeline to make it convenient for us to post our built Atom installers/packages somewhere other than Packagecloud.io
> 
> I vote for finding some other way to host the packages. I suppose we could distribute via personal repositories (PPA on Ubuntu, AUR on Arch, etc. etc.) and otherwise just post the installers/packages on GitHub Releases pages.
> 

We should use the release page of GitHub for now. Once we have our releases there it is very easy to make chocolatey, apt, homebrew packages from those. See https://github.com/Thilas/chocolatey-packages/tree/master/atom for example.

If we need to update the VSTs so they are uploaded to the release page we should do that.

**Comment by aminya** [comment]
Code signing is disabled now. But we run into issues in MacOS tests:
https://dev.azure.com/atomcommunity/atomcommunity/_build/results?buildId=29&view=logs&j=f6ee1f1b-a2ad-50b8-2639-9caf85d14d0f&t=d25d2602-b802-56cf-4ef2-197edda2fa63&l=239
```
15 warnings generated.
Downloading ripgrep failed: Error: Request failed: 403
    at ClientRequest.https.get.response (/Users/runner/runners/2.171.1/work/1/s/node_modules/vscode-ripgrep/lib/download.js:106:24)
    at Object.onceWrapper (events.js:286:20)
    at ClientRequest.emit (events.js:198:13)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:565:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:111:17)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! vscode-ripgrep@1.5.7 postinstall: `node ./lib/postinstall.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the vscode-ripgrep@1.5.7 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/runner/.atom/.apm/_logs/2020-07-04T07_06_31_678Z-debug.log[39m
child_process.js:642
    throw err;
    ^

Error: Command failed: /Users/runner/runners/2.171.1/work/1/s/apm/node_modules/atom-package-manager/bin/apm ci
    at checkExecSyncError (child_process.js:621:11)
    at Object.execFileSync (child_process.js:639:15)
    at module.exports (/Users/runner/runners/2.171.1/work/1/s/script/lib/run-apm-install.js:14:16)
    at Object.<anonymous> (/Users/runner/runners/2.171.1/work/1/s/script/bootstrap:44:1)
    at Module._compile (internal/modules/cjs/loader.js:959:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Function.Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1047:10)
    at internal/main/run_main_module.js:17:11 {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 50289,
  stdout: null,
  stderr: null
}

##[error]Bash exited with code '1'.
##[section]Finishing: Bootstrap build environment
```

**Comment by DeeDeeG** [comment]
Something went wrong during `npm install`, and it said the download cache had been corrupted. Can you hit "Retry failed runs"?

If it fails again the same way we would want to look into deleting the `node_modules` cache and starting fresh.

**Comment by DeeDeeG** [comment]
@aminya we just need to generate a GitHub personal access token and set that as an environment variable "GITHUB_TOKEN" for the pipelines. See comment above for details.

_Deleting this comment in a sec, just posting so you get a notification/email about it._

**Comment by aminya** [comment]
Thank you! I added the tokens. Let's see how it goes.
https://github.com/atom-ide-community/atom/runs/838906502

**Comment by aminya** [comment]
We get some warnings in the restore module_chace step:
https://dev.azure.com/atomcommunity/atomcommunity/_build/results?buildId=32&view=results

I am not sure how important they are.

**Comment by DeeDeeG** [comment]
I didn't realize upstream already uses the `GITHUB_TOKEN` environment variable properly, so it can in fact be a "secret variable"! Great! Now I have less worry in case Azure gets hacked! :laughing: 

We should really document all these variables and how to set them up properly, in case anyone else ever wants to fork Atom, or even just get CI running for their own personal use before submitting PRs, etc.

**Comment by aminya** [comment]
Now it almost gets to the end. There are some random tests on Windows:
https://dev.azure.com/atomcommunity/atomcommunity/_build/results?buildId=33&view=logs&j=4fbced83-508e-5fe0-c978-5c71ec0fc506&t=0cce7676-f298-5a2d-7f62-c11c27855377&l=2448

**Comment by DeeDeeG** [comment]
Just a flaky test. Timeouts are usually just the CI instance (or the code?) being slow. Not broken, but slow for some reason. I see you're re-running it, I expect it to pass.

It already passed on the other arch of Windows. I have some passing test runs on my personal fork's CI.

I think we basically have it working for testing the app. Not automated publishing, but CI does build the installers, so we could manually download those and post them in a (manually created) GitHub release as needed, for now. Figuring out automated releases is a good thing to do as well. But we could also start doing some code changes/try some PRs from upstream.

**Comment by aminya** [comment]
Yes. I wanted to make CI working before we merge something here. We need to publish one release for each change we make. I think this is almost ready. After the tests hopefully passed, I will merge #10, and we are ready to go 🚀

**Comment by DeeDeeG** [comment]
There are two ways to invite new users to Azure DevOps. I will document them here:

#### Invite in Azure DevOps Organization Settings

- Visit https://dev.azure.com/${your_organization_here}, e.g. https://dev.azure.com/atomcommunity/
- Press "Organization Settings" in the bottom corner (there is a gear icon)
- Press "Users" in the side nav area
- Press the "Add Users" button (it is a small blue button for me)
- Type the user's username or email
- Press the "Add to projects" dropdown and select the one you are running your pipelines in
- Press the "Add" button (a small blue button for me)

#### Invite from Azure DevOps Project Overview

- Visit https://dev.azure.com/${organization}/${project}, e.g. https://dev.azure.com/atomcommunity/atomcommunity
- Press the "Invite" button in the top corner (it is a small blue button for me)
- Type the username or email in the input box
- Press the "Add" button (it is a small blue button)

**Comment by aminya** [comment]
Any clue about why the windows tests take so much longer compared to MacOS? 

Windows uses windows-run.js which calls test script
https://github.com/atom-ide-community/atom/blob/master/script/vsts/windows-run.js
https://github.com/atom-ide-community/atom/blob/master/script/vsts/platforms/windows.yml#L122-L138

MacOS runs test script directly:
https://github.com/atom-ide-community/atom/blob/master/script/vsts/platforms/macos.yml#L181-L192

**Comment by DeeDeeG** [comment]
Here is the reasoning for windows-run.js, though I don't fully understand it: https://github.com/atom/atom/pull/18048

Something about needing 32-bit Node.js to build 32-bit Atom on Windows.

**Comment by aminya** [comment]
It seems to be related to https://github.com/microsoft/azure-pipelines-tasks/issues/6671.

> It shouldn't be hard to download Node and add it to the $PATH. So it feels like that has to be an unnecessarily slow/complicated workaround they chose.

If we can do this it is much better.

> I notice Windows x86 CI is much faster than Windows x64.

That's because it does not run the tests.

**Comment by aminya** [comment]
I made a PR to run the tests directly. https://github.com/atom-ide-community/atom/pull/31. Let's see if that is faster.

**Comment by aminya** [comment]
@DeeDeeG The cache is not preserved between different CI runs in different PRs. See #31. They may become on master.

**Comment by DeeDeeG** [comment]
Sadly, a lot of early work on this fork is going to include the CI config files. They are hashed as part of the cache match process. If the CI config is different, it's not considered a matching state of the repository, and new caching is required.

I could allow it to fall back on all the other matches, but I'm not sure I would propose it upstream...

This behavior (CI config changes require new cache) was first added here (https://github.com/atom/atom/pull/19504) so that things like updating Node in the CI config triggers bootstrapping.

**Comment by aminya** [comment]
I see. That is OK. If it works for other types of PRs that would be great. It makes sense to build from scratch when the CI config changes

**Comment by aminya** [comment]
@DeeDeeG Shouldn't we only cache the bootstrap when we are on the master branch? Other forks should be only able to restore our master cache, they should not overwrite the master cache.

**Comment by DeeDeeG** [comment]
There is a pretty clever system in place for that already.

If we get a PR from a fork, the cache there has READ access from the target branch at our repo AND READ access from `master` (if those are two different branches; usually the target will be `master`), and gets WRITE access to its own separate pile of caches just for that PR.
Forks can't overwrite our caches, generally speaking.

https://docs.microsoft.com/azure/devops/pipelines/release/caching?view=azure-devops#cache-isolation-and-security

That's why having good (and stable) caches in `master` will help is do quicker runs in the future. So long as we stop messing with the CI config files eventually. A lot of PRs are going to mess with the `package.json` and `package-lock.json` files too... Think of all the cache misses... :cry: Anyhow, it's about as good as it can be, I suppose.

**Comment by aminya** [comment]
> That's why having good (and stable) caches in `master` will help is do quicker runs in the future. So long as we stop messing with the CI config files eventually. A lot of PRs are going to mess with the `package.json` and `package-lock.json` files too... Think of all the cache misses... 😢 Anyhow, it's about as good as it can be, I suppose.

We should restore the cache even if we change package.json. `npm install` is smart enough to not redo everything from scratch!

**Comment by DeeDeeG** [comment]
In my experience, if there is anything in `node_modules` when you start an `npm install`, no native code will be built in the packages. Even if you Ctrl+C in the middle of the native code compilation.

I have had some of the weirdest, most difficult to diagnose issues from doing `npm install` over a partially/corruptly installed `node_modules`. Maybe that makes me paranoid and over-reacting. I suppose I would want to investigate this more. Ways it can fail and get good at recognizing those so we aren't confused if it happens. And best-case scenario we get better at proving nothing went wrong / good problem solving and deduction skills etc. etc.

With all this CI stuff taking up so much time, I think it might be time to just go for some upstream PRs and stuff like that. Real app maintenance. And... development. Stuff of that nature.

By the way I found some small things off with the pipelines that I'd like to correct, but I might have to pause the old pipelines and replace them, if that would be okay.

**Comment by aminya** [comment]
> By the way I found some small things off with the pipelines that I'd like to correct, but I might have to pause the old pipelines and replace them, if that would be okay.

Make pull requests and we will review. Don't pause the pipelines. Azure will queue yours automatically.

**Comment by aminya** [comment]
Azure apparently allows refactoring the repeated YAML configs using templates. If we can refactor our caching stuff that would be great:
```yaml
  - template: platforms/cache-windows.yml
```

**Comment by DeeDeeG** [comment]
I saw that docs page some time ago. I didn't have in mind to make these radical changes in case upstream does something and we have a mountain of delicate rebases to do.

But I am all for trying it.

I am tired because I have been staying up late. I should get some sleep. But yeah, I'll take that on. Would be glad to see what works. There is a lot of repeated stuff.

**Comment by aminya** [comment]
> I didn't have in mind to make these radical changes in case upstream does something and we have a mountain of delicate rebases to do.

This should not be a burden for us. We might apply a lot of changes as we move forward.

**Comment by DeeDeeG** [comment]
I said it before, I think we might have different visions of how hard a fork we want to make. I'm interested in this project, but I always want to contribute to upstream as well. I am not motivated so much by a desire to reinvent Atom as to keep it updating. But in that sense I can provide any technical skill I have to keeping some form of Atom updating, whether it's this or upstream... Preferably both!

Philosophy aside, yeah let's try things. I want to see what can be done. It's easy to change our approach as time goes on if need be, and this is still a bit improvisational.

**Comment by aminya** [comment]
We like to enhance upstream, but the main reason we created this fork was to have more freedom. Upstream is very slow in reviewing or reluctant in accepting the changes. We want to have our releases that the control is in the hands of the community and not the upstream team, which now only has few developers.

If you are only interested in minor changes, you might better work on your personal fork that tracks upstream. That will keep your PRs clean for them.

**Comment by aminya** [comment]
@DeeDeeG Let me know if you can try this. https://github.com/atom-ide-community/atom/issues/1#issuecomment-656485638
I will put #31 and #33 on hold until we refactor cache stuff.

We might close the upstream PR for #30 and instead make a PR for the refactored version.

**Comment by DeeDeeG** [comment]
I looked at this and I am trying to read the docs how the template paths are. Like if you can do a relative path `../file_in_parent_dir.yml`.

I also think moving _just_ the cache step out to its own template, one template per OS, would be more complecity for not much better readability (maybe worse due to having to track down and read two files). I am trying to think of a way to simplify with templates, not make it more complex, and I'm not sure. Maybe I'm overthinking. (I'm reluctant to try and be wrong and stuck fighting with CI for an hour...)

So if you want to try something I'll take a look at it. If you'd like me to post something with just the cache in a template I'll try it.

**Comment by aminya** [comment]
I made #38

**Comment by aminya** [comment]
@DeeDeeG Please see https://github.com/DeeDeeG/atom/pull/6. Edit your previous PR to upstream to include the new change. You might use `git cherry-pick` to only bring the latest commit.

**Comment by DeeDeeG** [comment]
I'd like to set the fetch-depth for the repository to 1. This fetches only the latest commit, and required objects, rather than the whole commit history and all of the associated objects of the entire history. Should speed up the initial repo fetch by about half.

Also, it might be silly, but caching the `node_modules` of the GetReleaseVersion step would eliminate most of the remaining time that step takes. GetReleaseVersion could finish instantly, and we'd be on to testing the actual environments, not just doing some metadata-collecting task.

Fetch-depth can be done in Azure DevOps UI or (I thin) in YAML. Whichever you prefer.

Caching the `node_modules` in GetReleaseVersion would have to change the YAML.

Let me know what you think.

(P.S. It's not much time savings, but it feels like a lot since it holds up the entire build. Remember, if you shave off two minutes ten times, you've shaved off twenty minutes. Small things are often the only attainable way of speeding up CI. Or at least they are the lower-hanging fruit, so to speak.)

**Comment by DeeDeeG** [comment]
I set the Pull Request to run every night on `master` (if it's been updated since that pipeline was last run on `master`), so pull request runs can have fewer cache misses.

I didn't want to trigger the Pull Request pipeline every time we merge into `master`, because we often need the runners for running actual PRs and for running the Release Branch Build pipeline on `master`, during those moments. I expect we'd cancel a lot of those. No problem if we want to change it, I just wanted to do something before we forget.

**Comment by DeeDeeG** [comment]
We can optimize the implementation of `Cache@2` by making sure that the bootstrap step is the last step of a [`job`].

(Basically split "bootstrap" and "build" into two `jobs`.)

This should cause the cache to be saved immediately after bootstrapping. It will then be available if building fails for whatever reason.

> [After all steps in the job have run and assuming a successful job status, a special "save cache" step is run for each "restore cache" step that was not skipped.](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#using-the-cache-task)

This eliminates the main downside of `Cache@2`, as we have implemented it, versus `SaveCache@1` as seen at upstream. (It is still more complex due to separate caches for each of the `node_modules` folders, but hopefully we can debug and manage that complexity.)

**Comment by aminya** [comment]
Caching a build that has failed does not make sense. There is a good reason behind this approach. We don't want to cache faulty configuration.

**Comment by DeeDeeG** [comment]
I respectfully disagree. There is more than one reason for a build failure, and only one of them is problems with dependencies. The project has its own code which can introduce errors/CI failures.

If the boostrap process finishes with no errors, it is a valid bootstrap result. It can be re-used for testing code tweaks that are outside of and a separate concern from the dependencies.

(If the dependencies in the `package.json`/`package-lock.json` files are updated, then the fingerprint of the dependencies change, which causes a bootstrap cache miss on the next run. Fresh bootstrapping is still done when needed.)

Recall that at the time I'm writing this, nothing from the build beyond the bootstrap process is saved.

Upstream have had this all figured out for some time, and to deviate from it (without providing an enhanced use of cachce of some kind to justify it) would introduce a regression in the availability of our cache without a concrete benefit.

**Comment by aminya** [comment]
This might sound possible in theory, but it does not work in practice. Other than having a cache that might not be able to build Atom we will waste a lot of time. 

Imagine we have a suitable bootstrap cache. It takes ~5min to prepare a system. Then we spend ~2-3min to restore the cache. Then the bootstrap step is slipped. Now we have wasted ~8min to fire up a system and restore a cache that is not used anywhere. Considering 4 operating systems, this becomes wasting half an hour (4*8 = 32min)

Now, if we instead use this 8min that we have spent so far to build atom, we save 30min for each CI run.

**Comment by DeeDeeG** [comment]
You may be right.

In terms of Cache@2: if there is a way to save the cache without having to close the Agent and spin up a new one, then that's more what I was thinking. I can't think of a way to do that at the moment, but I am keeping a close eye on the documentation that I read in case something makes this possible.

It is what upstream was doing with SaveCache@1, though. I think SaveCache@1 might be a bit slower to read/write the cache, by a very small amount, but I like that you can determine exactly when the save occurs. I am not going to start a long discussion about that here, because it's not a huge thing. But I have come to prefer the old caching strategy by a bit.

As a minor point to clear up (some build time numbers): I don't think the 4*8 minutes figure is the relevant amount of time, because we do not wait for runs serially, those times are in parallel. We do not experience the wait that way. So, back to parallel times: Indeed, 8 minutes * 2 for a total of 16 minutes is still a meaningful amount of time. So adding that time to builds would be bad. I agree this is bad, just I think the numbers are not as extreme as one hour.

**Comment by DeeDeeG** [comment]
You may be right.

In terms of Cache@2: if there is a way to save the cache without having to close the Agent and spin up a new one, then that's more what I was thinking. I can't think of a way to do that at the moment, but I am keeping a close eye on the documentation that I read in case something makes this possible.

It is what upstream was doing with SaveCache@1, though. I think SaveCache@1 might be a bit slower to read/write the cache, by a very small amount, but I like that you can determine exactly when the save occurs. I am not going to start a long discussion about that here, because it's not a huge thing. But I have come to prefer the old caching strategy by a bit.

As a minor point to clear up (some build time numbers): I don't think the 4*8 minutes figure is the relevant amount of time, because we do not wait for runs serially, those times are in parallel. We do not experience the wait that way. So, back to parallel times: Indeed, 8 minutes * 2 for a total of 16 minutes is still a meaningful amount of time. So adding that time to builds would be bad. I agree this is bad, just I think the numbers are not as extreme as one hour.

**Comment by DeeDeeG** [comment]
Now that #46 is merged, and the node/npm install parts of CI aren't in `platforms/$(os)`, we should probably add `platforms/templates/preparation.yml` and `bootstrap.yml` to the cache identifier. 

This is for the hypothetical case where we have updated Node, NPM, or some environment variables or config relevant to how the bootstrap and build should proceed.

**Comment by DeeDeeG** [comment]
CI isn't passing on `master`. Hmm. Not sure why.

**Comment by aminya** [comment]
> CI isn't passing on `master`. Hmm. Not sure why.

This is fixed in #63

**Comment by DeeDeeG** [comment]
Doing `npm install` is slower on Windows, due to NTFS's poor performance writing lots of small files versus EXT4.

Interesting discussion here: `https://github.com/rust-lang/rustup/issues/1540`

If possible/if it's okay with you, I'd like to revert https://github.com/atom-ide-community/atom/pull/3 (or explicitly set to use Linux) for the `GetReleaseVersion` and `Release` CI jobs, and use forward-slashes for cross-platform/Linux compatibility: https://github.com/DeeDeeG/atom/commit/fbc37428a847ea0977c9d0050688194720e49453

**Comment by aminya** [comment]
That sounds like a PR. We will not revert things on master anymore. It is OK if you want to do it in your PR.

**Comment by DeeDeeG** [comment]
Linux Build in CI has been failing by exiting out after the two package formats are apparently successfully built.

I also had a similar experience outside of CI on my personal machine, where the `dpkg-deb` command building the .deb package errors out.

I think this is flakiness, not a hard "100% of the time" issue, but it's still weird.

**Comment by DeeDeeG** [comment]
Also, just a heads up that there are some more hard-coded URLs pointing to `github.com/atom/atom` in various parts of the repo, and semi-relatedly, if we fall behind with which tags are on our `atom-nightly-releases` repo vs upstream's `atom-nightly-relleases` repo, it can cause the Windows build to fail on building differential/"update from previous version" partial .nupkg files.

I do think if we fix the hard-coded URLs like we did in #70, then this build failure scenario should also go away when using non-hard-coded URLs pointing to our own repos.

**Comment by aminya** [comment]
I will close this as this is mostly done.
